src/gordon_gcp/plugins/enricher.py

```diff
@@ -152,32 +161,49 @@ def __init__(self, config, http_client, success_channel, error_channel):
                 await asyncio.sleep(backoff)
                 backoff = backoff ** 2
 
-        msg = (f'Could not get external IP for {resource_name}: '
+        msg = (f'Could not get necessary information for {resource_name}: '
                f'{exception.__class__.__name__}: {exception}')
         raise exceptions.GCPGordonError(msg)
 
-    async def _create_A_rrecord(self, event_data, instance_data, msg_logger):
-        dns_zone = self.config['dns_zone']
-        hostname = event_data['resourceName'].split('/')[-1]
-        access_configs = instance_data['networkInterfaces'][0]['accessConfigs']
-        external_ip = access_configs[0]['natIP']
-
-        name = '.'.join([hostname, dns_zone])
-        msg_logger.debug(f'Creating A record: {name} -> {external_ip}')
+    def _create_A_rrecord(self, fqdn, external_ip, default_ttl, msg_logger):
+        msg_logger.debug(f'Creating A record: {fqdn} -> {external_ip}')
         return {
-            'name': name,
+            'name': fqdn,
             'rrdatas': [external_ip],
             'type': 'A',
-            'ttl': self.config['default_ttl']
+            'ttl': default_ttl
         }
 
-    async def _create_rrecords(self, event_data, msg_logger):
-        instance_data = await self._poll_for_instance_data(
-            event_data['resourceName'], msg_logger)
+    def _get_hostname(self, resource_name, msg_logger):
+        # pulled out for easy subclassing
+        return resource_name.split('/')[-1]
 
-        a_record = await self._create_A_rrecord(event_data, instance_data,
-                                                msg_logger)
-        return [a_record]
+    def _get_fqdn(self, resource_name, msg_logger):
+        # pulled out for easy subclassing
+        dns_zone = self.config['dns_zone']
+        hostname = self._get_hostname(resource_name)
+        return '.'.join([hostname, dns_zone])
+
+    def _get_external_ip(self, instance_data, msg_logger):
+        # pulled out for easy subclassing
+        interfaces = instance_data['networkInterfaces']
+        return interfaces[0]['accessConfigs'][0]['natIP']
+
+    def _get_internal_ip(self, instance_data, msg_logger):
+        # pulled out for easy subclassing
+        interfaces = instance_data['networkInterfaces']
+        return interfaces[0]['networkIP']
+
+    async def _create_rrecords(self, event_data, instance_data, msg_logger):
+        # async in case anything gets added in a subclass that needs to be async
+        fqdn = self._get_fqdn(event_data['resourceName'], msg_logger)
```

Comment:
Why not pull name from `instance_data['name']` instead?  Wherever we get the unqualified hostname from, I think passing it in instead could simplify the `_get_fqdn` function, but if we use instance data as the source we could erase the need for the for the `_get_hostname` method as well.

I realize this didn't change in this PR, but I think it becomes more obvious it's a little unintuitive when it has to get passed through several methods. 

@matt0bi Do you know if there was a specific reason for parsing from event resource name, rather than from the instance data? .As I understand it, they should be the same.

Reply:
@descent3me I don't remember any specific reason. Looking at the code, your suggestion is spot on!